src/Shared/ServerInfrastructure/StringUtilities.cs

```diff
@@ -387,28 +402,54 @@ public static string ConcatAsHexSuffix(string str, char separator, uint number)
                 length += str.Length;
             }
 
-            return string.Create(length, (str, separator, number), (buffer, tuple) =>
+            return string.Create(length, (str, separator, number), s_populateSpanWithHexSuffix);
+        }
+
+        private static void PopulateSpanWithHexSuffix(Span<char> buffer, (string str, char separator, uint number) tuple)
+        {
+            var (tupleStr, tupleSeparator, tupleNumber) = tuple;
+
+            var i = 0;
+            if (tupleStr != null)
             {
-                var (tupleStr, tupleSeparator, tupleNumber) = tuple;
-                char[] encode16Chars = s_encode16Chars;
+                tupleStr.AsSpan().CopyTo(buffer);
+                i = tupleStr.Length;
+            }
 
-                var i = 0;
-                if (tupleStr != null)
-                {
-                    tupleStr.AsSpan().CopyTo(buffer);
-                    i = tupleStr.Length;
-                }
+            buffer[i] = tupleSeparator;
+            i++;
 
-                buffer[i + 8] = encode16Chars[tupleNumber & 0xF];
-                buffer[i + 7] = encode16Chars[(tupleNumber >> 4) & 0xF];
-                buffer[i + 6] = encode16Chars[(tupleNumber >> 8) & 0xF];
-                buffer[i + 5] = encode16Chars[(tupleNumber >> 12) & 0xF];
-                buffer[i + 4] = encode16Chars[(tupleNumber >> 16) & 0xF];
-                buffer[i + 3] = encode16Chars[(tupleNumber >> 20) & 0xF];
-                buffer[i + 2] = encode16Chars[(tupleNumber >> 24) & 0xF];
-                buffer[i + 1] = encode16Chars[(tupleNumber >> 28) & 0xF];
-                buffer[i] = tupleSeparator;
-            });
+            if (Ssse3.IsSupported)
+            {
+                var lowNibbles = Ssse3.Shuffle(Vector128.CreateScalar(tupleNumber).AsByte(),s_shuffleMask);
+                var highNibbles = Sse2.ShiftRightLogical(Sse2.ShiftRightLogical128BitLane(lowNibbles, 2).AsInt32(), 4).AsByte();
+                var indices = Sse2.And(Sse2.Or(lowNibbles, highNibbles), Vector128.Create((byte)0xF));
+                // Lookup the hex values at the positions of the indices
+                var hex = Ssse3.Shuffle(s_asciiUpperCase, indices);
+                // The high bytes (0x00) of the chars have also been converted to ascii hex '0', so clear them out.
+                hex = Sse2.And(hex, Vector128.Create((ushort)0xFF).AsByte());
+
+                // This generates much more efficient asm than fixing the buffer and using
+                // Sse2.Store((byte*)(p + i), chars.AsByte());
+                Unsafe.WriteUnaligned(
+                    ref Unsafe.As<char, byte>(
+                        ref Unsafe.Add(ref MemoryMarshal.GetReference(buffer), i)),
+                    hex);
+            }
+            else
+            {
+                var hexEncodeMap = HexEncodeMap;
+                var number = (int)tupleNumber;
+
+                buffer[i + 7] = (char)hexEncodeMap[number & 0xF];
```

Comment:
The index for `buffer` isn't a constant, so I'm not sure if the JIT elides the subsequent bound checks.
Maybe it's worth to slice the buffer from `ì`, the the indices became constants.